repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/keystonev2/authorization/AuthorizationFilterTest.groovy

```diff
@@ -0,0 +1,200 @@
+        identityEndpoint = deproxy.addEndpoint(
+            properties.identityPort,
+            'identity service',
+            null,
+            fakeIdentityV2Service.handler)
+    }
+
+    def setup() {
+        fakeIdentityV2Service.resetDefaultParameters()
+        fakeIdentityV2Service.resetHandlers()
+        fakeIdentityV2Service.client_tenantid = UUID.randomUUID().toString()
+        fakeIdentityV2Service.client_token = UUID.randomUUID().toString()
+        fakeIdentityV2Service.originServicePort = properties.targetPort
+        reposeLogSearch.cleanLog()
+    }
+
+    def "User without valid service endpoint Port should receive a 403 FORBIDDEN response "() {
+        given: "IdentityService is configured with allowed endpoints that will differ from the user's requested endpoint"
+        fakeIdentityV2Service.originServicePort = 99999
+
+        when: "User sends a request through repose"
+        MessageChain mc = deproxy.makeRequest(
+            url: reposeEndpoint + "/v2/ss",
+            method: 'GET',
+            headers: [
+                'X-Auth-Token': fakeIdentityV2Service.client_token,
+                'X-Tenant-ID': fakeIdentityV2Service.client_tenantid])
+
+        then: "User should receive a 403 FORBIDDEN response"
+        mc.receivedResponse.code as Integer == SC_FORBIDDEN
+
+        and: "The reason should have been logged"
+        def foundLogs = reposeLogSearch.searchByString("User did not have the required endpoint")
+        foundLogs.size() == 1
+
+        and: "It should have never made it to the origin service"
+        mc.handlings.size() == 0
+    }
+
+    def "User without valid service endpoint URL should receive a 403 FORBIDDEN response "() {
+        given: "IdentityService is configured with allowed endpoints that will differ from the user's requested endpoint"
+        fakeIdentityV2Service.endpointUrl = "invalidurl"
+
+        when: "User sends a request through repose"
+        MessageChain mc = deproxy.makeRequest(
+            url: reposeEndpoint + "/v2/ss",
+            method: 'GET',
+            headers: [
+                'X-Auth-Token': fakeIdentityV2Service.client_token,
+                'X-Tenant-ID': fakeIdentityV2Service.client_tenantid])
+
+        then: "User should receive a 403 FORBIDDEN response"
+        mc.receivedResponse.code as Integer == SC_FORBIDDEN
+
+        and: "The reason should have been logged"
+        def foundLogs = reposeLogSearch.searchByString("User did not have the required endpoint")
+        foundLogs.size() == 1
+
+        and: "It should have never made it to the origin service"
+        mc.handlings.size() == 0
+    }
+
+    @Unroll
+    def "User sends the prefixed tenant (#tenantPrefix) in the URI should succeed and be extracted"() {
+        when: "User sends a request through repose"
+        MessageChain mc = deproxy.makeRequest(
+            url: reposeEndpoint + "/v2/extract/" + tenantPrefix + fakeIdentityV2Service.client_tenantid + "/ss",
```

Comment:
It took me a weirdly long time to figure out why this test was failing, but I think it's this test, as written, is invalid. Prefixes are stripped from tenants in the access response, not from the tenant we extract from the request.

Here are the current tests proving that for the KeystoneV2 filter:
https://github.com/rackerlabs/repose/blob/master/repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/keystonev2/tenantvalidation/StripTenantPrefixTest.groovy#L62-L96

The fix would be to set one of the tenant IDs in the Identity response to the prefixed value, and drop the prefix in the URI. Easy enough to do with:
```
fakeIdentityV2Service.client_tenantid = tenantPrefix + fakeIdentityV2Service.client_tenantid
```

Reply:
That makes sense. I guess I need more remedial Keystone v2 training. :smile: